### PR TITLE
Added Interfaces to some of the OData Client Library classes to Support Mocking 

### PR DESCRIPTION
**Pull request by ElizabethOkerio**
### Issues

*This pull request fixes issue #1250.*
### Description
*Added interfaces to some of the classes used by the OData code generator tool to support mocking*
**The classes affected are:**
1. DataServiceContext
2. DataServiceQuery
3. DataServiceQueryOfT
4. DataServiceQuerySingleOfT

**The interfaces added are:**
1. IDataServiceContext
2. IDataServiceQuery  
3. IDataServiceQueryOfT
4. IDataServiceQuerySingle

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.OData.Client/IDataServiceQueryOfT.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading.Tasks;
+
+namespace Microsoft.OData.Client
+{
+    public interface IDataServiceQueryOfT<TElement>
```
Comment (odero) [inline src/Microsoft.OData.Client/IDataServiceQueryOfT.cs]:
```suggestion
    public interface IDataServiceQuery<TElement>
```

File: src/Microsoft.OData.Client/DataServiceQueryOfT.cs [hunk 2]
```diff
@@ -20,7 +20,7 @@ namespace Microsoft.OData.Client
     /// </summary>
     /// <typeparam name="TElement">type of object to materialize</typeparam>
     [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1710:IdentifiersShouldHaveCorrectSuffix", Justification = "required for this feature")]
-    public class DataServiceQuery<TElement> : DataServiceQuery, IQueryable<TElement>
+    public class DataServiceQuery<TElement> : DataServiceQuery, IQueryable<TElement>, IDataServiceQueryOfT<TElement>
```
Comment (odero) [inline src/Microsoft.OData.Client/DataServiceQueryOfT.cs]:
```suggestion
    public class DataServiceQuery<TElement> : DataServiceQuery, IQueryable<TElement>, IDataServiceQuery<TElement>
```

File: src/Microsoft.OData.Client/IDataServiceQueryOfT.cs (hunk 1 above)
Comment (paulodero) [reply src/Microsoft.OData.Client/IDataServiceQueryOfT.cs]: The interfaces are to be used by the developer to test his code as opposed to unit tests in Odata client libraries.

**Comment by xuzhg** [comment]
> using System;

pay attention to the code style:

1) every C# file should have the copyright
2) Comments for all public classes, public properties, public methods..
3) ....

for detail, check the odata onenote.

---
Refers to: src/Microsoft.OData.Client/IDataServiceContext.cs:1 in 4bbea35. [](commit_id = 4bbea359d74649de91ff82483c22ace316848ec7, deletion_comment = False)

File: src/Microsoft.OData.Client/IDataServiceContext.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        EntityTracker EntityTracker { get; set; }
+        DataServiceClientFormat Format { get; }
+        bool IgnoreResourceNotFoundException { get; set; }
+        ReadOnlyCollection<LinkDescriptor> Links { get; }
+        ODataProtocolVersion MaxProtocolVersion { get; }
+        MergeOption MergeOption { get; set; }
+        Func<string, Uri> ResolveEntitySet { get; set; }
+        Func<Type, string> ResolveName { get; set; }
+        Func<string, Type> ResolveType { get; set; }
+        SaveChangesOptions SaveChangesDefaultOptions { get; set; }
```
Comment (xuzhg) [inline src/Microsoft.OData.Client/IDataServiceContext.cs]: I don't like to add an interface and add all public members into the interface.

Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/IDataServiceContext.cs]: Yes will restructure the interfaces.

File: src/Microsoft.OData.Client/DataServiceContext.cs [hunk 4]
```diff
@@ -50,7 +50,7 @@ internal enum UndeclaredPropertyBehavior
     /// The <see cref="T:Microsoft.OData.Client.DataServiceContext" /> represents the runtime context of the data service.
     /// </summary>
     [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Maintainability", "CA1506", Justification = "Central class of the API, likely to have many cross-references")]
-    public class DataServiceContext
+    public class DataServiceContext : IDataServiceContext
```
Comment (xuzhg) [inline src/Microsoft.OData.Client/DataServiceContext.cs]: >DataServiceContext : IDataServiceContext [](start = 17, length = 40)

So, how to test the "Mock" in your scenario?

Please add the Unit test, you can use "Moq" library in the unit test.

File: src/Microsoft.OData.Client/Microsoft.OData.Client.csproj [hunk 5]
```diff
@@ -449,6 +453,18 @@
...
+      <Link>IDataServiceQuerySingle.cs</Link>
+    </Compile>
+    <Compile Include="$(EnlistmentRoot)\src\Microsoft.OData.Client\IDataServiceQueryOfT.cs">
+      <Link>IDataServiceQueryOfT.cs</Link>
+    </Compile>
+    <Compile Include="$(EnlistmentRoot)\src\Microsoft.OData.Client\IDataServiceQuery.cs">
+      <Link>IDataServiceQuery.cs</Link>
+    </Compile>
+    <Compile Include="$(EnlistmentRoot)\src\Microsoft.OData.Client\IDataServiceContext.cs">
+      <Link>IDataServiceContext.cs</Link>
```
Comment (xuzhg) [inline src/Microsoft.OData.Client/Microsoft.OData.Client.csproj]: You add the interfaces into "Compile" in this file, why do you add them again and use the "link"?

File: src/Microsoft.OData.Client/DataServiceContext.cs (hunk 4 above)
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/DataServiceContext.cs]: Ok will add

**Comment by ElizabethOkerio** [comment]
I decided to make the public methods virtual instead of adding the interfaces to support mocking as it provided a cleaner way of doing it. Therefore am closing this PR and opening a new PR with the new changes.

**Pull request closed without merging**